File: 2914-find-the-safest-path-in-a-grid/find-the-safest-path-in-a-grid.cpp
class Solution {
public:
    int maximumSafenessFactor(vector<vector<int>>& grid) {
        int n = grid.size();

        queue<pair<int, int>> q;
        vector<vector<int>> safety(n, vector<int>(n, -1));

        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == 1) {
                    safety[i][j] = 0;
                    q.push({i, j});
                }
            }
        }
        vector<pair<int, int>> dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        while (!q.empty()) {
            auto [r, c] = q.front();
            q.pop();

            for (auto [dr, dc] : dirs) {
                int nr = r + dr;
                int nc = c + dc;

                if (nr >= 0 && nr < n && nc >= 0 && nc < n &&
                    safety[nr][nc] == -1) {
                    safety[nr][nc] = safety[r][c] + 1;
                    q.push({nr, nc});
                }
            }
        }

        priority_queue<tuple<int, int, int>> maxHeap;
        vector<vector<bool>> visited(n, vector<bool>(n, false));

        maxHeap.push({safety[0][0], 0, 0});
        visited[0][0] = true;

        while (!maxHeap.empty()) {
            auto [sf, r, c] = maxHeap.top();
            maxHeap.pop();

            if (r == n - 1 && c == n - 1) {
                return sf;
            }

            for (auto [dr, dc] : dirs) {
                int nr = r + dr;
                int nc = c + dc;

                if (nr >= 0 && nr < n && nc >= 0 && nc < n &&
                    !visited[nr][nc]) {
                    visited[nr][nc] = true;
                    int newSf = min(sf, safety[nr][nc]);
                    maxHeap.push({newSf, nr, nc});
                }
            }
        }
        return 0;
    }
};

File: 2914-find-the-safest-path-in-a-grid/find-the-safest-path-in-a-grid.py
class Solution:
    def maximumSafenessFactor(self, grid: List[List[int]]) -> int:
        n = len(grid)

        q = deque()
        safety = [[-1] * n for _ in range(n)]

        for i in range(n):
            for j in range(n):
                if grid[i][j] == 1:
                    safety[i][j] = 0
                    q.append((i, j))

        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        while q:
            r, c = q.popleft()

            for dr, dc in dirs:
                nr, nc = r + dr, c + dc

                if (0 <= nr < n) and (0 <= nc < n) and (safety[nr][nc] == -1):
                    safety[nr][nc] = safety[r][c] + 1
                    q.append((nr, nc))

        max_heap = [(-safety[0][0], 0, 0)]
        visited = [[False] * n for _ in range(n)]
        visited[0][0] = True

        while max_heap:
            sf, r, c = heapq.heappop(max_heap)
            sf = -sf

            if (r == n - 1) and (c == n - 1):
                return sf

            for dr, dc in dirs:
                nr, nc = r + dr, c + dc

                if (0 <= nr < n) and (0 <= nc < n) and not visited[nr][nc]:
                    visited[nr][nc] = True
                    new_sf = min(sf, safety[nr][nc])
                    heapq.heappush(max_heap, (-new_sf, nr, nc))